### PR TITLE
1.0 oneToMany join - Foreign Key joiner for KTables

### PR DESCRIPTION
**Pull request by bellemare**
Joins two KTables on the right table's foreign key projection. This allows you to avoid costly aggregations and groupByKeys, and the associated pains with that approach.

Current implementation supports all key types, provided the Serdes exist for them.

There are still a few TODOs in this project, which will be addressed during the next phase of optimization. I have tested this using the built in unit tests, but also using the Flyer Eventification application's unit tests, in staging and with production data volumes.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -361,7 +361,8 @@ subprojects {
     // version 7.x requires Java 8
     toolVersion = '6.19'
   }
-  test.dependsOn('checkstyleMain', 'checkstyleTest')
+  // Bellemare - re-enable the below
```
Comment (bellemare) [inline build.gradle:364]: Haha yes... I will re-add the style guide checks. It currently fails on a number of things.

File: doBuildAndRelease.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+rm -rf core/build/distributions/kafka_2.12-1.0.2-flipp.tgz
+gradle releaseTarGz -PscalaVersion=2.12
+rm -rf kafka_2.12-1.0.2-flipp/
+tar -xvzf core/build/distributions/kafka_2.12-1.0.2-flipp.tgz
+
+rm -rf ~/.m2/repository/org/apache/kafka/kafka-streams/1.0.2-flipp/kafka-streams-1.0.2-flipp.jar
+
+mkdir -p ~/.m2/repository/org/apache/kafka/kafka-streams/1.0.2-flipp/
+cp kafka_2.12-1.0.2-flipp/libs/kafka-streams-1.0.2-flipp.jar ~/.m2/repository/org/apache/kafka/kafka-streams/1.0.2-flipp/kafka-streams-1.0.2-flipp.jar
```
Comment (bellemare) [inline doBuildAndRelease.sh:10]: Utility script that I will remove from final pull

File: gradle.properties [hunk 3]
```diff
@@ -16,7 +16,7 @@
 group=org.apache.kafka
 # NOTE: When you change this version number, you should also make sure to update
 # the version numbers in tests/kafkatest/__init__.py and kafka-merge-pr.py.
-version=1.0.3-SNAPSHOT
+version=1.0.3-flipp
```
Comment (bellemare) [inline gradle.properties]: We're going to be maintaining this in-house in jfrog, until I can get this accepted into core Kafka (by handling the TODOs, further optimizations and having that pass the kafka code review)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStore.java [hunk 4]
```diff
@@ -98,6 +98,11 @@ public V get(final K key) {
         return new DelegatingPeekingKeyValueIterator<>(storeName, new CompositeKeyValueIterator<>(stores.iterator(), nextIteratorFunction));
     }
 
+    @Override
+    public KeyValueIterator<K, V> prefixScan(K prefix) {
+        throw new UnsupportedOperationException(getClass().getSimpleName() + " can't perform a prefix scan");
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStore.java:103]: this violates LSP, this is not a proper subclass for ReadOnlyKeyValueStore. Better to create a new interface that contains `KeyValueIterator<K, V> prefixScan(K prefix);` and then types (key value stores) that actually support this operation implement this interface.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java [hunk 5]
```diff
@@ -130,4 +130,9 @@ protected void removed(K key) {
     public KeyValueIterator<K, V> all() {
         return this.inner.all();
     }
+
+    @Override
+    public KeyValueIterator<K, V> prefixScan(K prefix) {
+        throw new UnsupportedOperationException(getClass().getSimpleName() + " can't perform a prefix scan");
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueLoggedStore.java:136]: this violates LSP, this is not a proper subclass for ReadOnlyKeyValueStore. Better to create a new interface that contains `KeyValueIterator<K, V> prefixScan(K prefix);` and then types (key value stores) that actually support this operation implement this interface.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java [hunk 6]
```diff
@@ -139,6 +139,11 @@ public synchronized V delete(K key) {
         return new DelegatingPeekingKeyValueIterator<>(name, new InMemoryKeyValueIterator<>(copy.entrySet().iterator()));
     }
 
+    @Override
+    public KeyValueIterator<K, V> prefixScan(K prefix) {
+        throw new UnsupportedOperationException(getClass().getSimpleName() + " can't perform a prefix scan");
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:144]: this violates LSP, this is not a proper subclass for ReadOnlyKeyValueStore. Better to create a new interface that contains `KeyValueIterator<K, V> prefixScan(K prefix);` and then types (key value stores) that actually support this operation implement this interface.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java [hunk 7]
```diff
@@ -184,6 +184,11 @@ public synchronized V delete(K key) {
         throw new UnsupportedOperationException("MemoryLRUCache does not support all() function.");
     }
 
+    @Override
+    public KeyValueIterator<K, V> prefixScan(K prefix) {
+        throw new UnsupportedOperationException(getClass().getSimpleName() + " can't perform a prefix scan");
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java:189]: this violates LSP, this is not a proper subclass for ReadOnlyKeyValueStore. Better to create a new interface that contains `KeyValueIterator<K, V> prefixScan(K prefix);` and then types (key value stores) that actually support this operation implement this interface.

File: streams/src/test/java/org/apache/kafka/test/NoOpReadOnlyStore.java [hunk 8]
```diff
@@ -60,6 +60,11 @@ public V get(final K key) {
         return null;
     }
 
+    @Override
+    public KeyValueIterator<K, V> prefixScan(K prefix) {
+        throw new UnsupportedOperationException(getClass().getSimpleName() + " can't perform a prefix scan");
```
Comment (kmatasfp) [inline streams/src/test/java/org/apache/kafka/test/NoOpReadOnlyStore.java:65]: this violates LSP, this is not a proper subclass for ReadOnlyKeyValueStore. Better to create a new interface that contains `KeyValueIterator<K, V> prefixScan(K prefix);` and then types (key value stores) that actually support this operation implement this interface.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 9]
```diff
@@ -582,4 +593,35 @@ public void onRestoreEnd(final TopicPartition topicPartition,
...
+
+        @Override
+        public synchronized boolean hasNext() {
+            if(!super.hasNext()){
+                return false;
+            }
+
+            byte[] rawNextKey = super.peekRawKey();
+            for (int i = 0; i < rawPrefix.length; i++) {
+                if (i == rawNextKey.length) {
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:615]: this check can be only done once outside of the loop

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableKTableRangeJoin.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+	private ValueJoiner<VL, VR, V> joiner;
+	private KTableRangeValueGetterSupplier<CombinedKey<KL,KR>,VR> right;
+
+    //Performs Left-driven updates (ie: new One, updates the Many).
+    public KTableKTableRangeJoin(KTableRangeValueGetterSupplier<CombinedKey<KL,KR>,VR> right,
+                                 ValueJoiner<VL, VR, V> joiner){
+
+    	this.right = right;
+        this.joiner = joiner;
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableKTableRangeJoin.java]: indent, final?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableJoinMergeProcessorSupplier.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+public class KTableJoinMergeProcessorSupplier<K0,V0,K,V,KO,VO> implements KTableProcessorSupplier<K0, V0, V0>{
+	private boolean sendOldValue = false;
+
+
+	private KTableValueGetterSupplier<K, V> leftValueGetter;
+	private KTableValueGetterSupplier<K0, VO> rightValueGetter;
+	private ValueMapper<K0, K>leftKeyExtractor;
+	private ValueJoiner<V, VO, V0> joiner;
+
+    public KTableJoinMergeProcessorSupplier(KTableValueGetterSupplier<K, V> leftValueGetter,
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableJoinMergeProcessorSupplier.java]: indent?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableJoinMergeProcessorSupplier.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.kafka.streams.kstream.internals.KTableValueGetterSupplier;
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+public class KTableJoinMergeProcessorSupplier<K0,V0,K,V,KO,VO> implements KTableProcessorSupplier<K0, V0, V0>{
+	private boolean sendOldValue = false;
+
+
+	private KTableValueGetterSupplier<K, V> leftValueGetter;
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableJoinMergeProcessorSupplier.java]: can these be final?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableKTableRangeJoin.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+	@Override
+    public Processor<KL, Change<VL>> get() {
+        return new KTableKTableJoinProcessor(right);
+    }
+	
+
+    private class KTableKTableJoinProcessor extends AbstractProcessor<KL, Change<VL>> {
+
+		private KTableRangeValueGetter<CombinedKey<KL,KR>,VR> rightValueGetter;
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableKTableRangeJoin.java]: indent?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableKTableRangeJoin.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    //Performs Left-driven updates (ie: new One, updates the Many).
+    public KTableKTableRangeJoin(KTableRangeValueGetterSupplier<CombinedKey<KL,KR>,VR> right,
+                                 ValueJoiner<VL, VR, V> joiner){
+
+    	this.right = right;
+        this.joiner = joiner;
+    }
+
+	@Override
```
Comment (kmatasfp) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/onetomany/KTableKTableRangeJoin.java]: indent?

**Comment by bellemare** [comment]
This PR is not quite finished yet - there are some things in here that will be TODOs, and not all concerns will be addressed in this PR. The goal is to have this working well-enough to go into production, with the caveat being that this is to be merged into core kafka in the future. This means that certain shortcuts have been taken, such as not reworking the entire Materialized API to limit rangescans to just rocksDB tables. For the purposes of our implementations there will simply be some limitations. That being said, I will update this with some better commenting and variable names, formatting, etc, but this was intended as a quick way to get some eyes on the code (as per our session yesterday).

**Pull request closed without merging**